### PR TITLE
[Dy2Static] Convert if stmt and move code

### PR DESCRIPTION
**Pull request by liym27**
### PR types: 
New features and Bug fixes

### PR changes: 
dygraph to static graph

#### 1. Add function `convert_ifelse` to run the transformed code dynamically:
- All if/else and ifExpr stmts are transformed whether it depends on Tensor or not;
- But in run-time, the code will be run in different ways (1) run paddle op like `cond`; (2) run python if/else. How it runs depends on whether it depends on Tensor.
#### 2. Fix bug: 
cast var dtype to bool in function _run_paddle_logical_XX 
#### 3. Code optimization: 
3.1 Add `LogicalTransformer` : move `LogicalTransformer` out from `LoopTransformer` and `IfElseTransformer`, and run `LogicalTransformer` separately.

3.2 Remove unnecessary class:  MergeAssignTransformer, NodeTestTransformer and IfConditionVisitor in ifelse_transformer.py

### CODE REVIEW

File: python/paddle/fluid/dygraph/dygraph_to_static/ifelse_transformer.py [hunk 1]
```diff
@@ -93,242 +84,18 @@ def visit_IfExp(self, node):
...
-            new_node_str = "fluid.layers.logical_not({})".format(arg)
-            # gast.parse returns Module(body=[expr(value=...)])
-            new_node = gast.parse(new_node_str).body[0].value
-            logic_tensor_name = unique_name.generate(LOGIC_NOT_PREFIX)
-            assign_name, assign_node = create_assign_node(logic_tensor_name,
-                                                          new_node)
-            self._new_assign_nodes.append(assign_node)
-            return assign_name
-
-        return node
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/ifelse_transformer.py:214]: Thanks to optimize these codes. Great work.

**Review by Aurelius84 — DISMISSED** [review]
LGTM

**Review by chenwhql — DISMISSED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/dygraph_to_static/yolov3.py [hunk 2]
```diff
@@ -314,19 +314,21 @@ def forward(self,
...
-                scores=yolo_scores,
-                score_threshold=cfg.valid_thresh,
-                nms_top_k=cfg.nms_topk,
-                keep_top_k=cfg.nms_posk,
-                nms_threshold=cfg.nms_thresh,
-                background_label=-1)
-            return pred
-        else:
-            return sum(self.losses)
+        # TODO(liym27): Uncomment code after "return" statement can be transformed correctly.
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/yolov3.py:317]: Why is this return commented? Convert_if shouldn't transform because self.is_train is not tensor.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py [hunk 3]
```diff
@@ -106,9 +106,13 @@ def forward(self, inputs, label=None):
             acc = fluid.layers.accuracy(input=x, label=label)
             loss = fluid.layers.cross_entropy(x, label)
             avg_loss = fluid.layers.mean(loss)
-            return x, acc, avg_loss
-        else:
-            return x
+
+        # TODO: Uncomment code after "return" statement can be transformed correctly.
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py:110]: Why is this return commented? Convert_if shouldn't transform because `if label is not None` is not tensor.

Comment (liym27) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py:110]: If not commented , this `forward` function is transformed as follows, but it run python if actually because the pred `label is not None` is not a Tensor.  `Return` statements needs to be transformed specially. I will do this in the future PR.
```Python
def forward(self, inputs, label=None):
    x = fluid.dygraph.dygraph_to_static.convert_call(self.inference)(inputs)

    def true_fn_0(label, x):
        acc = fluid.layers.accuracy(input=x, label=label)
        loss = fluid.layers.cross_entropy(x, label)
        avg_loss = fluid.layers.mean(loss)
        return x, acc, avg_loss
        return

    def false_fn_0(x):
        return x
        return
    fluid.dygraph.dygraph_to_static.convert_operators.convert_ifelse(label
         is not None, lambda : fluid.dygraph.dygraph_to_static.convert_call
        (true_fn_0)(label, x), lambda : fluid.dygraph.dygraph_to_static.
        convert_call(false_fn_0)(x))
```

After commented, the transformed code is 
```Python
def forward(self, inputs, label=None):
    x = fluid.dygraph.dygraph_to_static.convert_call(self.inference)(inputs)
    acc = (fluid.dygraph.dygraph_to_static.variable_trans_func.
        data_layer_not_check(name='acc', shape=[-1], dtype='float32'))
    avg_loss = (fluid.dygraph.dygraph_to_static.variable_trans_func.
        data_layer_not_check(name='avg_loss', shape=[-1], dtype='float32'))

    def true_fn_0(acc, avg_loss, label, x):
        acc = fluid.layers.accuracy(input=x, label=label)
        loss = fluid.layers.cross_entropy(x, label)
        avg_loss = fluid.layers.mean(loss)
        return acc, avg_loss

    def false_fn_0(acc, avg_loss):
        return acc, avg_loss
    acc, avg_loss = (fluid.dygraph.dygraph_to_static.convert_operators.
        convert_ifelse(label is not None, lambda : fluid.dygraph.
        dygraph_to_static.convert_call(true_fn_0)(acc, avg_loss, label, x),
        lambda : fluid.dygraph.dygraph_to_static.convert_call(false_fn_0)(
        acc, avg_loss)))
    return x, acc, avg_loss


```

File: python/paddle/fluid/tests/unittests/dygraph_to_static/yolov3.py (hunk 2 above)
Comment (liym27) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/yolov3.py:317]: The same as bolow

Comment (zhhsplendid) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/yolov3.py:317]: Discussed offline, we will add it back after we support early return.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py (hunk 3 above)
Comment (zhhsplendid) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py:110]: Discussed offline, we will add it back after we support early return.

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by Aurelius84 — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]
